packages/libs/error-stack/src/ser.rs

```diff
@@ -0,0 +1,152 @@
+//! Implementation of general report serialization.
+//!
+//! The value can be of any type, currently only printable attachments and context are supported, in
+//! the near future any values will be supported through the use of hooks.
+//!
+//! The returned JSON returned is a list of all current sources with the following output:
+//!
+//! ```json
+//! {
+//!     "context": "string value",
+//!     "attachments": ["all attachments leading up to this context"],
+//!     "sources": [] // recursive render using `frame.sources()`
+//! }
+//! ```
+
+use alloc::{format, vec, vec::Vec};
+
+use serde::{ser::SerializeMap, Serialize, Serializer};
+
+use crate::{AttachmentKind, Context, Frame, FrameKind, Report};
+
+struct SerializeAttachment<'a>(&'a Frame);
+
+impl<'a> Serialize for SerializeAttachment<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let Self(frame) = self;
+
+        #[allow(clippy::match_same_arms)]
+        match frame.kind() {
+            FrameKind::Context(_) => {
+                // `SerializeContext` ensures that no context is ever serialized
+                unreachable!()
+            }
+            FrameKind::Attachment(AttachmentKind::Opaque(_)) => {
+                // TODO: for now opaque attachments are unsupported, upcoming PR will fix that
+                // `SerializeContext` ensures that no such attachment is added
+                unreachable!()
+            }
+            FrameKind::Attachment(AttachmentKind::Printable(attachment)) => {
+                format!("{attachment}").serialize(serializer)
+            }
+        }
+    }
+}
+
+struct SerializeAttachments<'a, 'b>(&'a [&'b Frame]);
+
+impl<'a, 'b> Serialize for SerializeAttachments<'a, 'b> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.collect_seq(
+            self.0
+                .iter()
+                .copied()
+                .filter(|attachment| {
+                    // for now opaque attachments are ignored
```

Comment:
Is this `unreachable!`?

Reply:
How so?

Reply:
clarified in [4c67564](https://github.com/hashintel/hash/pull/1290/commits/4c675645442c693dbd0877bef178c065b48bc2ff)

Reply:
Thank you 🙂